Enable tensor fragments for zero 2 & 3 (#2727)
microsoft/DeepSpeed: docs/code-docs/source/zero3.rst

.. autoclass:: deepspeed.zero.TiledLinear
    :members:

Debugging
---------

> Debugging ZeRO training is complicated by the partitioning of parameters, gradients, and optimizer states.
> To simplify debugging, DeepSpeed provides the following routines for accessing individual model states
> in their full form (i.e., un-partitioned).

Reviewer: ```suggestion
Debugging ZeRO training is complicated by the partitioning of parameters, gradients, and optimizer states.
None of these 3 groups of parameters can be normally accessed because of that. To overcome that DeepSpeed provides the following utilities for accessing individual model states in their unpartitioned form.

Please note that these utilities must be called by all processes, even if you decide to do something with the result only in the main process. If all processes don't participate these utilities will hang waiting for all processes to send their contribution.
```

Reviewer: please review/extend my additions on when the data is valid.

.. autofunction:: deepspeed.utils.safe_get_full_fp32_param

.. autofunction:: deepspeed.utils.safe_get_full_grad

.. autofunction:: deepspeed.utils.safe_get_full_optimizer_state

> These routines can be used in a training loop as shown in the following snippet.

Reviewer: ah, you prefer "routines" instead of "utilities" - that works, let's then rename these back int he paras above, to keep things consistent.

.. code-block:: python

    backward(loss)
    [...]
    from deepspeed.utils import safe_get_full_fp32_param, safe_get_full_grad, safe_get_full_optimizer_state
    for n, lp in model.named_parameters():

>         # 1. grad lookup must be called after `backward` and before `step` for z2

Reviewer: ```suggestion
        # 1. grad lookup must be called after `backward` and before `step` for zero2
```

Reviewer: and for zero3?

.. autoclass:: deepspeed.zero.TiledLinear
    :members:

Debugging
---------

> Debugging ZeRO training is complicated by the partitioning of parameters, gradients, and optimizer states.
> None of these 3 groups of parameters can be normally accessed because of that. To overcome that DeepSpeed provides the following utilities for accessing individual model states in their unpartitioned form.
> 
> Please note that routines utilities must be called by all processes, even if you decide to do something with the result only in the main process. If all processes don't participate these utilities will hang waiting for all processes to send their contribution.

Reviewer: ```suggestion
Debugging ZeRO training is complicated by the partitioning of parameters, gradients, and optimizer states.
None of these 3 groups of tensors (model states) can be normally accessed because of that. To overcome that DeepSpeed provides the following routines for accessing individual model states in their unpartitioned form.

Important: Please note that these utilities must be called by all processes participating in the training, even if you decide to do something with the result only in the main process. If all processes don't participate these utilities will hang waiting for all processes to send their contribution.

Additionally, you must be aware that these routines return correct data only in specific phases of the training. So for examples the gradients are valid after ``backward`` and before ``step``. The optimizer states are updated after ``step``. Same goes for fp32 master weights.
```

Reviewer: sorry, I was still editing the suggestion - I reloaded it - now it's good to go.
